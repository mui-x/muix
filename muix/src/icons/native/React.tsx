import createSvgIcon from './lib/create-svg-icon'

export default createSvgIcon(
  'react',
  'React',
  true
)