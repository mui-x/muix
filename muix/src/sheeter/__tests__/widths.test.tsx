/** @jsx platform.createElement */

import React from 'react'

import {
  platform, atomizeRuleset, toClassNamesWithQuery,
  $width, setActWidth, useWidths, $WidthsQuery
} from "reactxx-sheeter"

import { initPlatform, dump, mount } from "./init-platform"

describe("SHEETER $WIDTHS", () => {

  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let ruleset

    it("01 empty", () => {
      ruleset = atomizeRuleset([
        { color: 'green' },
        $width([640, 1024], null, {}, undefined),
      ])
      expect(ruleset).toMatchSnapshot()
    })

    it("02 simple", () => {
      ruleset = atomizeRuleset([
        $width([640, 1024], { color: 'red' }),
      ])
      expect(ruleset).toMatchSnapshot()
    })

    describe("03 toClassNamesWithQuery's width change", () => {
      beforeEach(() => initPlatform(isWeb))

      const rulesets = [
        $width([0, 640], { color: 'red' }),
        $width([640, 1024], { color: 'green' }),
        $width(1024, { color: 'blue' }),
      ]
      !window.isWeb && it("01: native, 300", () =>
        dump(toClassNamesWithQuery<$WidthsQuery>({ $widths: { actWidth: 300 } }, ...rulesets))
      )
      !window.isWeb && it("02: native, 640", () =>
        dump(toClassNamesWithQuery<$WidthsQuery>({ $widths: { actWidth: 640 } }, ...rulesets))
      )
      !window.isWeb && it("03: native, 1024", () =>
        dump(toClassNamesWithQuery<$WidthsQuery>({ $widths: { actWidth: 1024 } }, ...rulesets))
      )
      !window.isWeb && it("04: native, undefined", () =>
        dump(toClassNamesWithQuery<$WidthsQuery>(undefined, ...rulesets))
      )
      window.isWeb && it("05: web", () =>
        dump(toClassNamesWithQuery<$WidthsQuery>(undefined, ...rulesets))
      )
    })

    describe("04 change width", () => {
      beforeEach(() => initPlatform(isWeb, {dataTraceFlag: 'short'}))

      const getSheet = () => atomizeRuleset([
        $width([0, 640], { color: 'red' }),
        $width([640, 1024], { color: 'green' }),
        $width(1024, { color: 'blue' }),
      ])

      const useApp = () => {
        const { actWidth, breakpoints, getWidthMap } = useWidths()
        const query: $WidthsQuery = {
          $widths: { actWidth, breakpoints }
        }
        const sheetRoot = React.useMemo(getSheet)
        // component code
        const renderCount = React.useRef(0)
        renderCount.current++

        const root = toClassNamesWithQuery(query, sheetRoot)

        return { getWidthMap, root, renderCount: renderCount.current }
      }

      const App: React.SFC = () => {
        const { root, renderCount } = useApp()
        return <div classNameX={root}>{`rendered: ${renderCount}x`}</div>
      }
      App['$c$'] = true

      const WidthsMapApp: React.SFC = () => {
        const { root, renderCount, getWidthMap } = useApp()
        const [mobile, tablet, desktop] = getWidthMap([640, 1024])
        return <div classNameX={root}>
          {`${mobile ? 'mobile' : ''}${tablet ? 'tablet' : ''}${desktop ? 'desktop' : ''} (rendered: ${renderCount}x)`}
        </div>
      }
      WidthsMapApp['$c$'] = true

      const test = (App: React.ComponentType) => {
        setActWidth(300)
        const wrapper = mount(<App />)
        expect(wrapper.container).toMatchSnapshot()
        setActWidth(640)
        expect(wrapper.container).toMatchSnapshot()
        setActWidth(1024)
        expect(wrapper.container).toMatchSnapshot()
        wrapper.unmount()
      }

      it("01 use just width styles", () => test(App))
      it("02 use width flags and styles", () => test(WidthsMapApp))

    })
  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})

