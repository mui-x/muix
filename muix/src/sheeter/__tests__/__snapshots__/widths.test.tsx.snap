// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHEETER $WIDTHS ## NATIVE ## 01 empty 1`] = `
Array [
  Object {
    "color": "green @[0]",
  },
]
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 02 simple 1`] = `
Array [
  Object {
    "color": "red @/$width",
  },
]
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 01: native, 300 1`] = `
Array [
  Object {
    "color": "red @/$width",
  },
]
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 01: native, 300 2`] = `
Object {
  "color": Object {
    "tracePath": "/$width",
    "value": "red",
  },
}
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 01: native, 300 3`] = `
Object {
  "color": "red",
}
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 02: native, 640 1`] = `
Array [
  Object {
    "color": "green @/$width",
  },
]
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 02: native, 640 2`] = `
Object {
  "color": Object {
    "tracePath": "/$width",
    "value": "green",
  },
}
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 02: native, 640 3`] = `
Object {
  "color": "green",
}
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 03: native, 1024 1`] = `
Array [
  Object {
    "color": "blue @/$width",
  },
]
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 03: native, 1024 2`] = `
Object {
  "color": Object {
    "tracePath": "/$width",
    "value": "blue",
  },
}
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 03: native, 1024 3`] = `
Object {
  "color": "blue",
}
`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 04: native, undefined 1`] = `null`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 04: native, undefined 2`] = `null`;

exports[`SHEETER $WIDTHS ## NATIVE ## 03 toClassNamesWithQuery's width change 04: native, undefined 3`] = `undefined`;

exports[`SHEETER $WIDTHS ## WEB ## 01 empty 1`] = `
Array [
  Array [
    ".a { color:green } @[0]",
  ],
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 02 simple 1`] = `
Array [
  Array [
    "@media (min-width: 640px) and (max-width: 1023px).a { color:red } @/$width",
  ],
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 01: native, 300 1`] = `
Array [
  Array [
    "@media (max-width: 639px).a { color:red } @/$width",
  ],
  Array [
    "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  ],
  Array [
    "@media (min-width: 1024px).c { color:blue } @/$width",
  ],
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 01: native, 300 2`] = `
Array [
  "@media (min-width: 1024px).c { color:blue } @/$width",
  "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  "@media (max-width: 639px).a { color:red } @/$width",
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 01: native, 300 3`] = `"c b a"`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 02: native, 640 1`] = `
Array [
  Array [
    "@media (max-width: 639px).a { color:red } @/$width",
  ],
  Array [
    "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  ],
  Array [
    "@media (min-width: 1024px).c { color:blue } @/$width",
  ],
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 02: native, 640 2`] = `
Array [
  "@media (min-width: 1024px).c { color:blue } @/$width",
  "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  "@media (max-width: 639px).a { color:red } @/$width",
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 02: native, 640 3`] = `"c b a"`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 03: native, 1024 1`] = `
Array [
  Array [
    "@media (max-width: 639px).a { color:red } @/$width",
  ],
  Array [
    "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  ],
  Array [
    "@media (min-width: 1024px).c { color:blue } @/$width",
  ],
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 03: native, 1024 2`] = `
Array [
  "@media (min-width: 1024px).c { color:blue } @/$width",
  "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  "@media (max-width: 639px).a { color:red } @/$width",
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 03: native, 1024 3`] = `"c b a"`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 04: native, undefined 1`] = `
Array [
  Array [
    "@media (max-width: 639px).a { color:red } @/$width",
  ],
  Array [
    "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  ],
  Array [
    "@media (min-width: 1024px).c { color:blue } @/$width",
  ],
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 04: native, undefined 2`] = `
Array [
  "@media (min-width: 1024px).c { color:blue } @/$width",
  "@media (min-width: 640px) and (max-width: 1023px).b { color:green } @/$width",
  "@media (max-width: 639px).a { color:red } @/$width",
]
`;

exports[`SHEETER $WIDTHS ## WEB ## 03 toClassNamesWithQuery's width change 04: native, undefined 3`] = `"c b a"`;
